### ISSUE TITLE
File#copy does not support files above 5 GB

### ISSUE THREAD

**Issue by stanhu** [fog/fog-aws#577]
`Fog::AWS::Storage::File#copy` uses the single CopyObject API: https://github.com/fog/fog-aws/blob/ab288f29a0974d64fd8290db41080e5578be9651/lib/fog/aws/models/storage/file.rb#L100-L105

However, this does not use multipart uploads. As a result, files above 5 GB fail to transfer.

In addition, I noticed the AWS SDK uses 10 threads in parallel to speed up the transfer (https://github.com/aws/aws-sdk-ruby/blob/d2dc0213758da0aee100e485f2dfb090ffa7dbd5/gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_file_uploader.rb#L133-L163). As a result, uploads via the SDK are significantly faster than fog-aws (e.g. 6 GB file takes 7 minutes as opposed to a minute with the `aws` CLI).

Certainly implementing a multi-threaded version that uses multipart uploads could be implemented, but is there a reason fog-aws doesn't just call out to the AWS SDK in the first place?

**Comment by geemus** [comment]
@stanhu hey, thanks for the detailed report and checking in. Copy was implemented like that, I think largely because it was simple and seemed good-enough at the time. We certainly could think about an implementation more like what you suggest. 

I worry a bit about the additional complexity of managing threads, debugging issues, etc if we did it ourselves. As for using aws-sdk directly, the short answer is that `aws-sdk` is not a dependency of fog-aws and we are trying to keep the dependency list small/short. If someone needs/wants this functionality, I guess I might suggest they utilize the aws-sdk directly, rather than adding it for this in particular. Do you think we should advise people to use aws-sdk directly if they need this, or is it important to change this and/or add this capability here?

**Comment by stanhu** [comment]
@geemus That's a good question. Probably documentation should suffice for now. However, setting aside the copy case, I was surprised by how dramatic basic upload performance was without multithreading for a 2 GB file:

|Case|Time|
|-----|----|
|fog-aws|57 s|
|aws CLI|9.6 s|

We use CarrierWave with fog-aws, fog-google, fog-rackspace, as well as our own fog-azure fork to support multiple object storage providers. CarrierWave uses the `copy_object` request (https://github.com/carrierwaveuploader/carrierwave/blob/2a0fefcd7762f91292fbc84911fc91f142699766/lib/carrierwave/storage/fog.rb#L453) to store a file. Switching to using the aws-sdk appears to be straightforward, but many of our tests expect to be testing against the Fog framework. Now we have to add a bit of complexity by treating AWS as a special case.

As another data point, fog-google appears to use the Google SDK directly, but it doesn't appear there is support for multithreaded uploads in the Ruby or even in the Go SDK. `gsutil` implements this parallel upload mode inside the application: https://github.com/GoogleCloudPlatform/gsutil/blob/4c634ad85dbd13c3445e2e96320c3d91db9ca73e/gslib/utils/copy_helper.py#L1055-L1278.

I did a quick spike test to see whether a single-threaded multipart transfer would help copy performance for a 2 GB file, but sadly this code below was still slower than the single PUT request (93 s). That's not too surprising given that we need multiple threads to see significant gains.

```diff
diff --git a/lib/fog/aws/models/storage/file.rb b/lib/fog/aws/models/storage/file.rb
index 15b1d3d..94af1f4 100644
--- a/lib/fog/aws/models/storage/file.rb
+++ b/lib/fog/aws/models/storage/file.rb
@@ -99,7 +99,14 @@ module Fog
         #
         def copy(target_directory_key, target_file_key, options = {})
           requires :directory, :key
-          service.copy_object(directory.key, key, target_directory_key, target_file_key, options)
+          if self.content_length > 0
+            src = key.delete_prefix('/')
+            upload_part_options = options.merge({ 'x-amz-copy-source' => "#{directory.key}/#{key}" })
+            multipart_copy(options, upload_part_options, target_directory_key, target_file_key)
+          else
+            service.copy_object(directory.key, key, target_directory_key, target_file_key, options)
+          end
+
           target_directory = service.directories.new(:key => target_directory_key)
           target_directory.files.head(target_file_key)
         end
@@ -294,6 +301,50 @@ module Fog
           service.complete_multipart_upload(directory.key, key, upload_id, part_tags)
         end
 
+        def multipart_copy(options, upload_part_options, target_directory_key, target_file_key)
+          # Initiate the upload
+          puts "saving to bucket #{target_directory_key} => #{target_file_key}"
+          res = service.initiate_multipart_upload(target_directory_key, target_file_key, options)
+          upload_id = res.body["UploadId"]
+
+          puts "upload ID is #{upload_id}"
+
+          # Store ETags of upload parts
+          part_tags = []
+
+          current_pos = 0
+          chunk_size = 10 * 1024 * 1024
+
+          # Upload each part
+          # TODO: optionally upload chunks in parallel using threads
+          # (may cause network performance problems with many small chunks)
+          while current_pos < self.content_length do
+            start_pos = current_pos
+            end_pos = [current_pos + chunk_size, self.content_length - 1].min
+            range = "bytes=#{start_pos}-#{end_pos}"
+            puts range
+
+            upload_part_options['x-amz-copy-source-range'] = range
+            part_upload = service.upload_part(target_directory_key, target_file_key, upload_id, part_tags.size + 1, '', upload_part_options)
+            etag = Nokogiri::XML(part_upload.body).css('ETag').first.text.gsub("\"", '')
+            part_tags << etag
+            current_pos = end_pos + 1
+          end
+
+#          if part_tags.empty? #it is an error to have a multipart upload with no parts
+#            part_upload = service.upload_part(directory.key, key, upload_id, 1, '', part_headers('', upload_part_options))
+#            part_tags << part_upload.headers["ETag"]
+#          end
+
+        rescue => e
+          # Abort the upload & reraise
+          service.abort_multipart_upload(target_directory_key, target_file_key, upload_id) if upload_id
+          raise
+        else
+          # Complete the upload
+          service.complete_multipart_upload(target_directory_key, target_file_key, upload_id, part_tags)
+        end
+
         def encryption_headers
           if encryption && encryption_key
             encryption_customer_key_headers
```
</details>

**Comment by geemus** [comment]
@stanhu thanks for the additional details. That time difference between them if they are both multithreaded and working more or less the same does seem surprising, I wonder what accounts for the difference then...

**Comment by geemus** [comment]
Er sorry, I meant if they are both single-threaded.

**Comment by stanhu** [comment]
@geemus There are more round-trips in the serial, multipart case, so that might be part of the issue. In any event, I've submitted https://github.com/fog/fog-aws/pull/578 to fix the 5GB+ copy case. I think making this multithreaded is fairly straightforward. Would you be open to a follow-up pull request for that?

**Comment by geemus** [comment]
@stanhu thanks. Yeah, I would expect some overhead from round-trips, but not nearly that much (unless there are really way more round trips than I thought). I'm open to the possibility of multithreaded support, but do worry about supporting/maintaining it (as my experience with multithreaded stuff has been that it is much more difficult to debug/etc if anything goes wrong, particularly when other folks are reporting the bugs).
